kernel/safe_typing.ml

```diff
@@ -1086,23 +1131,18 @@ let build_module_body params restype senv =
 let allow_delayed_constants = ref false
 
 let propagate_senv newdef newenv newresolver senv oldsenv =
-  let now_cst, later_cst = List.partition Future.is_val senv.future_cst in
   (* This asserts that after Paral-ITP, standard vo compilation is behaving
    * exctly as before: the same universe constraints are added to modules *)
-  if not !allow_delayed_constants && later_cst <> [] then
+  if not !allow_delayed_constants && not (HandleMap.is_empty senv.future_cst) then
     CErrors.anomaly ~label:"safe_typing"
       Pp.(str "True Future.t were created for opaque constants even if -async-proofs is off");
   { oldsenv with
     env = newenv;
     modresolver = newresolver;
     revstruct = newdef::oldsenv.revstruct;
     modlabels = Label.Set.add (fst newdef) oldsenv.modlabels;
-    univ =
-      List.fold_left (fun acc cst ->
-        Univ.ContextSet.union acc (Future.force cst))
-      (Univ.ContextSet.union senv.univ oldsenv.univ)
-      now_cst;
-    future_cst = later_cst @ oldsenv.future_cst;
+    univ = (Univ.ContextSet.union senv.univ oldsenv.univ);
+    future_cst = HandleMap.fold HandleMap.add senv.future_cst oldsenv.future_cst;
```

Comment:
Why don't we just 
```suggestion
    future_cst = senv.future_cst;
```
here?

If that's incorrect why do we fold add instead of union?

Reply:
To be fair I just adapted the code without much thought. I have no idea why we have not two, but *three* different safe environments in this function and what invariants relate them. As for `union` it's just that fold + add seems more natural to me that the weird union API for maps, but this is essentially a mater of taste and there is no serious reason otherwise.

Reply:
This function comes from here
https://github.com/coq/coq/commit/fac949450909b5ef17078f220ae809cf54ae3f08#diff-24b536f3006f7ced4984e883e87dbe8a0a1b0b6c921bf661f7f307c41bd88d53R558

Reply:
If I read the code correctly, this function is actually buggy in master. Namely, joining the environment inside a module is not really joining it since the delayed constraints from outside the module being currently defined are not added to the global environment. This bug transposes in this PR with an anomaly when joining from withing a module. Small reproducible example:
```
Lemma foo : True.
Proof.
constructor.
Qed.

Module Foo.
(* Joining here results in 'Anomaly "Missing opaque with identifier 0"' *)
```
I am not sure that claiming we're correct while silently ignoring universe constraints is a good behaviour either, but at least it's less visible to the end user. I'll fix this with an additional patch.

Reply:
Well, in some sense it is not a bug, since you can't save a vo file with an open module. But I prefer the old behavior (no failure, no guarantee) to the new anomaly. Can't you turn it into a better error?

Reply:
> Well, in some sense it is not a bug

This further weakens the expected invariant of joining the environment. I thought it was "things make sense just right after you joined the env", but it seems you also need to add a bunch of provisos like "no modules are opened" and "the age of the captain is 42". I don't mind trying to be more lenient, but at the very least we need an actual spec.

Reply:
Can't we just get joining to work in modules? Is there more to do than add `future_cst = senv.future_cst;` in start_module/modtype?

Reply:
I don't think so, that's what I've been starting writing.

Reply:
>  "no modules are opened" and "the age of the captain is 42".

I guess you could also make it work with an open module, but it is more work for very little in return IMO.
Turning the anomaly into "sorry, you can't join the env with an open module" is much simpler and user friendly.

Reply:
It's trivial now, I've just added a patch for that. That said I do not manage to get a working `.fake` test for that situation, see the last commit (0355b359594f3be02b67e5c0bafd13696acfafe4). It fails with
```
...
FAIL 0 [Universe inconsistency. Cannot enforce v < u because u < v.]
fake_ide: error: Universe inconsistency. Cannot enforce v < u because u < v.
The command failed as expected
fake_ide: error: No state found
```
That's produced by `FAILJOIN` so it recognizes that a failure is expected but then it behaves weirdly w.r.t. the expected state.

Reply:
Hum, IMO the new join does not get attached (by the STM?) a last_good_id, so it backtracks to the initial state or something like that. I can have a look. What happens in CoqIDE when you hit the wheels button?

Reply:
It displays an error in the right pane, but does not backtrack anywhere. Any subsequent join fails as well.

Reply:
```
gares@ollypat:~/COQ/move-opaque-out-of-kernel$ test-suite/ide/fake_ide.exe _build/install/default/bin/coqidetop.opt test-suite/ide/join-module2.fake 
Running: _build/install/default/bin/coqidetop.opt --xml_format=Ppcmds
Add ((" Universes u v. ",-1),(1,true))
GOOD (2,(Inl ,""))
Add ((" Lemma foo : True. ",-2),(2,true))
GOOD (3,(Inl ,""))
Add ((" Proof. ",-3),(3,true))
GOOD (4,(Inl ,""))
Add ((" refine (let _ : Type@{v} := Type@{u} in _). ",-4),(4,true))
GOOD (5,(Inl ,""))
Add ((" constructor. ",-5),(5,true))
GOOD (6,(Inl ,""))
Add ((" Qed. ",-6),(6,true))
GOOD (7,(Inl ,""))
Add ((" Lemma bar : True. ",-7),(7,true))
GOOD (8,(Inl ,""))
Add ((" Proof. ",-8),(8,true))
GOOD (9,(Inl ,""))
Add ((" refine (let _ : Type@{u} := Type@{v} in _). ",-9),(9,true))
GOOD (10,(Inl ,""))
Add ((" constructor. ",-10),(10,true))
GOOD (11,(Inl ,""))
Add ((" Qed. ",-11),(11,true))
GOOD (12,(Inl ,""))
Add ((" Module Foo. ",-12),(12,true))
GOOD (13,(Inl ,""))
Status true
FAIL 9 (1,43)[The term "Type" has type "Type@{v+1}" while it is expected to have type
 "Type@{u}" (universe inconsistency: Cannot enforce v < u because u < v).]
fake_ide: error: The term "Type" has type "Type@{v+1}"
                 while it is expected to have type "Type@{u}"
                 (universe inconsistency: Cannot enforce v < u because u
                 < v).
The command failed as expected
Edit_at 9
GOOD Inl 

*[          ,   2] Universes u v. 
*[          ,   3] Lemma foo : True. 
*[          ,   4] Proof. 
*[          ,   5] refine (let _ : ...
*[          ,   6] constructor. 
*[          ,   7] Qed. 
*[          ,   8] Lemma bar : True. 
*[          ,   9] Proof. 
Quitting fake_ide
gares@ollypat:~/COQ/move-opaque-out-of-kernel$ echo $?
0
```
How did you get the error?

Reply:
As you can see, I get a `Fail 9`, which means go back to state 9, the last known good one. And it seems to go there. CoqIDE should backtrack there as well.

Maybe you did fix the bug without realizing it?

Reply:
Running your command works with the same output as yours, but running the test-suite still fails on this example. What's going on?

Reply:
Running your command works with the same output as yours, but running the test-suite still fails on this example. What's going on?

Reply:
@gares any update on this?

Reply:
No sorry, I had no time to look into this. I'll do it now.

Reply:
So
```
(cd test-suite/; ide/fake_ide.exe '/home/gares/COQ/move-opaque-out-of-kernel'/_build/install/default/bin/coqidetop.opt "ide/join-module2.fake" "-q -async-proofs on -async-proofs-tactic-error-resilience off -async-proofs-command-error-resilience off ")
```
fails! Apparently `-async-proofs off` is the default for fake_ide (and not for CoqIDE). Which is silly² since the whole point of fake_ide is to test CoqIDE as it is... so the bug is real.

Reply:
> so the bug is real.

That's a net progress, but this does not explain the bug, unfortunately...

Reply:
```
Status true
FAIL 0 [Universe inconsistency. Cannot enforce v < u because u < v.]
fake_ide: error: Universe inconsistency. Cannot enforce v < u because u < v.
The command failed as expected
fake_ide: error: No state found
```
As I was saying, this means that the error does not carry a `safe_id`, 0 is the dummy id. So jumping back to 0 is an error. The STM was passing a `fix_exn` argument to `Future.*` in order to enrich any exception escaping the thunk with this piece of info. I don't know how this is handled now, since I have some memory of this argument going away at some point. I'll check it, but if this ring a bell please tell me.

Reply:
Here:
```
        let cert0 = Future.join cert in
        let () = Global.fill_opaque cert0 in
```
You force the thunk, which hopefully would raise exceptions enriched the right way. But the you check outside the thing that `cert0` is ok, so IMO this check does not raise a correctly wrapped exception. (Maybe the bug was already there). Does this sound like a plausible explanation?

Reply:
I see, so that probably means I have to wrap the exception coming from the join operation with the function provided with the corresponding future in vernac/opaques. I'll try that.

Reply:
If I'm right, `Future.chain cert (fun cert0 -> Global.fill_opaque cert0)` should do it...

Reply:
Indeed, something like that. It's a bit ugly to have to work around the Future API but for the time being that'll do it.

Reply:
It seems to have worked, miraculously...